services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py

```diff
@@ -84,72 +148,48 @@ async def _sequentially_upsert_tasks_from_project(
         published_nodes: List[NodeID],
         str_project_uuid: str,
     ) -> None:
-        # start by removing the old tasks if they exist
-        await self.connection.execute(
-            sa.delete(comp_tasks).where(comp_tasks.c.project_id == str(project.uuid))
-        )
-        # create the tasks
 
-        for internal_id, node_id in enumerate(project.workbench, 1):
-            node: Node = project.workbench[node_id]
+        # NOTE: really do an upsert here because of issue https://github.com/ITISFoundation/osparc-simcore/issues/2125
 
-            service_key_version = ServiceKeyVersion(
-                key=node.key,
-                version=node.version,
-            )
-            node_class = to_node_class(service_key_version.key)
-            node_details: ServiceDockerData = None
-            node_extras: ServiceExtras = None
-            if node_class == NodeClass.FRONTEND:
-                node_details = _get_fake_service_details(service_key_version)
-            else:
-                node_details = await director_client.get_service_details(
-                    service_key_version
-                )
-                node_extras: ServiceExtras = await director_client.get_service_extras(
-                    service_key_version
-                )
-            if not node_details:
-                continue
-
-            requires_mpi = False
-            requires_gpu = False
-            if node_extras:
-                requires_gpu = NodeRequirement.GPU in node_extras.node_requirements
-                requires_mpi = NodeRequirement.MPI in node_extras.node_requirements
-            image = Image(
-                name=service_key_version.key,
-                tag=service_key_version.version,
-                requires_gpu=requires_gpu,
-                requires_mpi=requires_mpi,
-            )
+        list_of_comp_tasks_in_project = await _generate_tasks_list_from_project(
+            project, director_client, published_nodes
+        )
 
-            task_db = CompTaskAtDB(
-                project_id=project.uuid,
-                node_id=node_id,
-                schema=NodeSchema(
-                    inputs=node_details.inputs, outputs=node_details.outputs
-                ),
-                inputs=node.inputs,
-                outputs=node.outputs,
-                image=image,
-                submit=datetime.utcnow(),
-                state=(
-                    RunningState.PUBLISHED
-                    if node_id in published_nodes
-                    and node_class == NodeClass.COMPUTATIONAL
-                    else RunningState.NOT_STARTED
-                ),
-                internal_id=internal_id,
-                node_class=node_class,
+        result: RowProxy = await self.connection.execute(
+            sa.select([comp_tasks.c.node_id]).where(
+                comp_tasks.c.project_id == str(project.uuid)
             )
+        )
+        node_id_to_delete = [
```

Comment:
what happens if the node you are about to delete happens to be running?

Reply:
The only running nodes are computational. there is a check in /routes/computations.py that checks for that. you cannot modify the comp_tasks table if the computational pipeline is running.

Reply:
actually this does not change. the upserting is now done rightfully and before it was done by re-creating the tasks.